hail/python/hailtop/aiogoogle/client/storage_client.py

```diff
@@ -0,0 +1,204 @@
+            **kwargs))
+        return InsertObjectStream(it, request_task)
+
+    async def get_object(self, bucket: str, name: str, **kwargs) -> GetObjectStream:
+        if 'params' in kwargs:
+            params = kwargs['params']
+        else:
+            params = {}
+            kwargs['params'] = params
+        assert 'alt' not in params
+        params['alt'] = 'media'
+
+        resp = await self._session.get(
+            f'https://storage.googleapis.com/storage/v1/b/{bucket}/o/{urllib.parse.quote(name, safe="")}', **kwargs)
+        return GetObjectStream(resp)
+
+    async def get_object_metadata(self, bucket: str, name: str, **kwargs) -> Any:
+        params = kwargs.get('params')
+        assert not params or 'alt' not in params
+        return await self.get(f'/b/{bucket}/o/{urllib.parse.quote(name, safe="")}', **kwargs)
+
+    async def delete_object(self, bucket: str, name: str, **kwargs) -> None:
+        await self.delete(f'/b/{bucket}/o/{urllib.parse.quote(name, safe="")}', **kwargs)
+
+    async def _list_objects(self, bucket: str, **kwargs) -> Any:
+        return await self.get(f'/b/{bucket}/o', **kwargs)
+
+
+class GoogleStorageAsyncFS(AsyncFS):
+    def __init__(self, storage_client: Optional[StorageClient] = None):
+        if not storage_client:
+            storage_client = StorageClient()
+        self._storage_client = storage_client
+
+    def schemes(self) -> Set[str]:
+        return {'gs'}
+
+    @staticmethod
+    def _get_bucket_name(url: str) -> Tuple[str, str]:
+        parsed = urllib.parse.urlparse(url)
+        if parsed.scheme != 'gs':
+            raise ValueError(f"invalid scheme, expected gs: {parsed.scheme}")
+
+        name = parsed.path
+        if name:
+            assert name[0] == '/'
+            name = name[1:]
+
+        return (parsed.netloc, name)
+
+    async def open(self, url: str, mode: str = 'r') -> AsyncStream:
+        if not all(c in 'rwxabt+' for c in mode):
+            raise ValueError(f"invalid mode: {mode}")
+        if 't' in mode and 'b' in mode:
+            raise ValueError(f"can't have text and binary mode at once: {mode}")
+        if 'b' not in mode:
+            raise ValueError(f"text mode not supported: {mode}")
+        if 'x' in mode:
+            raise ValueError(f"exclusive creation not supported: {mode}")
+        if 'a' in mode:
+            raise ValueError(f"append mode not supported: {mode}")
+        if '+' in mode:
+            raise ValueError(f"updating not supported: {mode}")
+        if ('r' in mode) + ('w' in mode) != 1:
+            raise ValueError(f"must have exactly one of read/write mode: {mode}")
+
+        bucket, name = self._get_bucket_name(url)
+
+        if 'r' in mode:
+            return await self._storage_client.get_object(bucket, name)
+
+        assert 'w' in mode
+        return await self._storage_client.insert_object(bucket, name)
+
+    async def mkdir(self, url: str) -> None:
```

Comment:
This is where I got stuck when I tried to do this. I'd think that if you call `mkdir` and then followed by `isdir` that the directory would exist. But I didn't want a bunch of hanging files that ended in slashes. Like you said, unifying the local file system and the google one with object storage is not straightforward and it's not worth trying to unify them.

Reply:
> I'd think that if you call mkdir and then followed by isdir that the directory would exist.

Yep, there were a few choices here, and I think the simplest is to actually break the mkdir/isdir relationship.  You notice in my isdir test, I both call mkdir and create a file before calling isdir.  The other option was not to include isdir, and I'm very tempted to do this.  Directories are not a real concept in gs://, and it is maybe a mistake to try to emulate them.

The other option is to rename isdir/mkdir.  What it `isdir(p)` really means: are there are objects with `p/` as a prefix, and `mkdir(p)` means: you can now create an object with prefix `p/`.

I think once we get the interface sorted (including directory listing), we should document how we're thinking about it very carefully.